### PR TITLE
TPP-MLIR as out-of-tree dependency

### PR DESCRIPTION
**Pull request by slyalin**
Based on https://github.com/rengolin/openvino/tree/enable-tpp-mlir. Add include directories to `transformations` and link dependencies to openvino library only.

### CODE REVIEW

File: src/common/transformations/src/transformations/mlir/mlir_op.cpp [hunk 1]
```diff
@@ -64,9 +72,9 @@ void prepareMLIRKernelWithoutWrapper(mlir::OwningOpRef<mlir::ModuleOp>& module)
     // A set of default passes that lower any input IR to LLVM
     PassManager pm(module->getContext());
 
-#if 0  // TODO: if TPP is available
+#if TPP_MLIR && false  // FIXME: Still cannot enable it: undefined reference to mlir::createConvertOpenMPToLLVMPass(), mlir::createParallelLoopToGpuPass() etc.
```
Comment (slyalin) [inline src/common/transformations/src/transformations/mlir/mlir_op.cpp]: @rengolin, looks like we need to add more libraries in this list: https://github.com/slyalin/openvino/blob/mlir/src/cmake/openvino.cmake#L45-L59. But I don't understand how quickly find out which ones. Could you help?

Comment (rengolin) [reply src/common/transformations/src/transformations/mlir/mlir_op.cpp]: They should be in the MLIR paths, they're set by CMake via tricks like `add_llvm_executable` ([Link](https://github.com/plaidml/tpp-mlir/blob/295409bfbb271f9d5ad08096213ca7efe7210491/tools/tpp-opt/CMakeLists.txt#L24)) and `add_mlir_library` ([link](https://github.com/plaidml/tpp-mlir/blob/295409bfbb271f9d5ad08096213ca7efe7210491/lib/TPP/CMakeLists.txt#L12C1-L12C17))

Comment (slyalin) [reply src/common/transformations/src/transformations/mlir/mlir_op.cpp]: These macros cannot be used in openvino I suppose, as they are adding a new library or executable target, but we already have the library. Did you mean I can reverse-engineer those macros and understand how to get the required libraries? Is there a more regular way to do that without adding a new target?

Comment (rengolin) [reply src/common/transformations/src/transformations/mlir/mlir_op.cpp]: Either way. If you reverse engineer, it should be just a matter of finding the header/lib paths for MLIR and add those to the targets you need.

Comment (slyalin) [reply src/common/transformations/src/transformations/mlir/mlir_op.cpp]: OK, I found a way to build it successfully. Just using `MLIR_ALL_LIBS` property. But enabling it in this place fails in runtime with segfault.

Comment (rengolin) [reply src/common/transformations/src/transformations/mlir/mlir_op.cpp]: Where in the code is the segfault? TPP-MLIR? MLIR? LLVM?

File: src/cmake/openvino.cmake [hunk 2]
```diff
@@ -41,29 +41,17 @@ target_include_directories(${TARGET_NAME} PUBLIC
...
 target_link_libraries(${TARGET_NAME} PRIVATE openvino::reference
                                              openvino::shape_inference
                                              openvino::pugixml
                                              ${CMAKE_DL_LIBS}
                                              Threads::Threads
-                                             ${MLIR_OPENVINO_LIBS})
+                                             ${MLIR_ALL_LIBS})
+
+add_tpp_mlir_libs(${TARGET_NAME})
+
```
Comment (rengolin) [inline src/cmake/openvino.cmake:54]:
```suggestion
add_tpp_mlir_includes(${TARGET_NAME})

```

File: src/common/transformations/src/transformations/mlir/mlir_op.cpp [hunk 3]
```diff
@@ -53,6 +53,14 @@
 #include "mlir/Target/LLVMIR/Export.h"
 #include "mlir/Target/LLVMIR/ModuleTranslation.h"
 
+#ifdef TPP_MLIR // If TPP is available
+#include "TPP/Dialect/Check/CheckDialect.h"
+#include "TPP/Dialect/Perf/PerfDialect.h"
+#include "TPP/Dialect/Xsmm/XsmmDialect.h"
+#include "TPP/GPU/Utils.h"
+#include "TPP/Passes.h"
```
Comment (rengolin) [inline src/common/transformations/src/transformations/mlir/mlir_op.cpp:58]:
```suggestion
#include "TPP/Passes.h"
#include "TPP/PassBundles.h"
```

File: cmake/tpp-mlir.cmake [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            TPPPipeline
+            TPPCheckDialect
+            TPPCheckToLoops
+            TPPGPU
+            TPPIR
+            TPPLinalgToFunc
+            TPPLinalgToXSMM
+            TPPPerfDialect
+            TPPPerfToFunc
+            TPPPerfToLoop
```
Comment (rengolin) [inline cmake/tpp-mlir.cmake:19]:
```suggestion
            TPPPerfToLoop
            TPPPipeline
            TPPPassBundles
```

**Review by rengolin — COMMENTED** [review]
The two changes I propose will make it compile but it's still missing the MLIR libraries. I'm trying to see if I can move this logic into TPP MLIR itself.

File: src/cmake/openvino.cmake (hunk 2 above)
Comment (slyalin) [reply src/cmake/openvino.cmake:54]: Shouldn't affect the build, as no code outside `transformations` use these headers. And the build of `transformations` uses this directive.

**Pull request merged**